Refactoring

RUMirNul opened the pull request:
Изменение конфигурации приложения 
Некоторые пересылаемые данные замены на DTO 
Приведение переменных/методов/классов к camelcase
Убрано return null

Commit:
Renaming

Commit:
Changed using Map on AllCurrencyDto in all methods and classes.
Replaced RestTemplate on WebClient in AllCurrencyService

Commit:
Changed using parameters in some methods
Replaced RestTemplate on WebClient in FromToCurrencyService
Generation bot message has been moved to separate methods

Commit:
Changed method of configuration application

Commit:
Minor fix

AndreyPlis commented on pom.xml:
```diff
@@ -29,6 +29,17 @@
+            <artifactId>spring-boot-starter-webflux</artifactId>
+        </dependency>
+
+        <dependency>
```
а зачем эта зависимость? должно и так работать

RUMirNul replied:
Без неё не работает WebClient

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/CurrencyController.java:
```diff
@@ -1,13 +1,15 @@
+import javax.naming.ServiceUnavailableException;
 import java.util.Map;
 import java.util.Set;
 import java.util.regex.Pattern;
```
аннотация контроллер не нужна тут больше подойдет @Service

RUMirNul replied:
Done

AndreyPlis commented:
1) в pom.xml перейди на 17 джаву и спринг 2.7
2) пакеты начни сначала разбиавать по слоям а не по фичам, то есть пакеты tgbot cuncurrencyapy апи на верхнем уровне не нужны, верхний уровень такой: config. model, repository, srvice,controller, exceptions and etc
3) добавь редми что сервис делать и как, желательно на английском
4) не хватает тестов, как проверить что все работает?

AndreyPlis commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,14 @@
+
+currency:
+  api:
+    allPath: https://api.getgeoapi.com/v2/currency/list?api_key=APITOKEN
```
1 не надо в ресурсы хардкодить фулл путь, только домен https://api.getgeoapi.com/ 
2 allPath мне не понятно название) имена должны быть понятны человеку который видет первый раз.  лучше длинное имя но понятное чем короткое и не понятное

RUMirNul replied:
1 Не очень понимаю как по другому, через билдер url в webClient? Тогда ссылка будет захардкодена в коде 
2 Done

AndreyPlis replied:
1 здесь укаывается только хост а в коде дописывается конкретный путь и парамтеры

RUMirNul replied:
Done

AndreyPlis commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,14 @@
+currency:
+  api:
+    allPath: https://api.getgeoapi.com/v2/currency/list?api_key=APITOKEN
+    convertPath: https://api.getgeoapi.com/v2/currency/convert?api_key=APITOKEN&from={from}&to={to}&amount={amount}&format=json
```
аналогично выше

RUMirNul replied:
Done

AndreyPlis commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,14 @@
+server:
+  port: 5000
+
+telegram:
```
не надо дробить уровни впустую! уровни дробятся только в том случае если на уровне могут быть значения
в твоем случае из 
telegram:
  bot:
превратится
telegram-bot:

правила именования в ямле штрих между словами

RUMirNul replied:
Done

AndreyPlis commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,14 @@
+    token: BOTTOKEN
+    webhookPath: https://d2c9-94-241-55-71.eu.ngrok.io
+
+currency:
```
все проперти определияемые в приложении заносятся с префиксом app

то есть должно быть так
app:
  telegram-bot:
.........
  cuncurrency-api:

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/CurrencyController.java:
```diff
@@ -23,31 +25,34 @@ public CurrencyController(AllCurrencyService allCurrencyService, FromToCurrencyS
 
-    public Map<String, ToCurrencyConvert> getCalcRateAmount(String amount, String from, String to) throws InputAmountException, IllegalArgumentException {
-        if (!Pattern.compile("\\d{1,13}([\\.\\,]\\d{1,5})?").matcher(amount).matches()) {
+    public FromToCurrency calculateRateAmount(ConvertParametersDto parameters) throws InputAmountException, IllegalArgumentException, ServiceUnavailableException {
```
зачем в сигнатуре явный проброс разных эксепшенов throws InputAmountException, IllegalArgumentException, ServiceUnavailableException ты их где проверяешь?
нет то throws Exception

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/CurrencyController.java:
```diff
@@ -23,31 +25,34 @@ public CurrencyController(AllCurrencyService allCurrencyService, FromToCurrencyS
-        if (!Pattern.compile("\\d{1,13}([\\.\\,]\\d{1,5})?").matcher(amount).matches()) {
+    public FromToCurrency calculateRateAmount(ConvertParametersDto parameters) throws InputAmountException, IllegalArgumentException, ServiceUnavailableException {
+
+        if (!Pattern.compile("\\d{1,13}([\\.\\,]\\d{1,5})?").matcher(parameters.getAmount()).matches()) {
```
Любые строки или числа выносятся в константы с читаемым именем чтто означает, в данном случае я не понимаю что значит этот паттерн нигде нет его описания)

AndreyPlis replied:
более того компиляция регксов затратная операция зачем делать компиляцию регекса на каждый раз? выглядит так что можно оптимизировать и сделать один раз

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/CurrencyController.java:
```diff
@@ -23,31 +25,34 @@ public CurrencyController(AllCurrencyService allCurrencyService, FromToCurrencyS
+        if (validateCurrencyCode(parameters.getFrom()) && validateCurrencyCode(parameters.getTo())) {
+            FromToCurrency fromToCurrency = fromToCurrencyService.calculateRateAmount(parameters);
+
+            if (fromToCurrency.getStatus().equalsIgnoreCase(ApiMessage.SUCCESS.getMessage())) {
```
equalsIgnoreCase тут не к месту  сделай в  fromToCurrency также енам и сравнивай енам с енамом

RUMirNul replied:
Не актуально, была переделана логика

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/CurrencyController.java:
```diff
@@ -23,31 +25,34 @@ public CurrencyController(AllCurrencyService allCurrencyService, FromToCurrencyS
+            if (fromToCurrency.getStatus().equalsIgnoreCase(ApiMessage.SUCCESS.getMessage())) {
+                return fromToCurrency;
+            } else {
+                throw new ServiceUnavailableException();
```
а почему? нет ни сообщения юзеру с кодом ошибки и эксепшен не удачен, почему сервис не доступен?  если ошибка внтури твоего сервиса на запрос  то контроллер отдает бедреквест в данном случае, или пятисотку в крайний случай но не как сервис не доступен. Сервис не доступен имеет абсолютно другой смысл,

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/CurrencyController.java:
```diff
@@ -23,31 +25,34 @@ public CurrencyController(AllCurrencyService allCurrencyService, FromToCurrencyS
             }
+
         } else {
             throw new IllegalArgumentException();
```
а почему? где какой то код ошибки или что произошло?

RUMirNul replied:
Подставил сообщение BotMessage.EXCEPTION_NO_VALID_CURRENCY_CODE_MESSAGE
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/CurrencyController.java:
```diff
@@ -23,31 +25,34 @@ public CurrencyController(AllCurrencyService allCurrencyService, FromToCurrencyS
-        if (!(allCurrencyNames.isEmpty() || allCurrencyNames == null)) {
+        Map<String, String> allCurrencyNames = allCurrencyService.getAllCurrency().getCurrencies();
+
+        if (!(allCurrencyNames == null || allCurrencyNames.isEmpty())) {
```
не надо проверять на нулл коллекцию лучше сделай так чтобы allCurrencyService возвращал пустую колекцию

AndreyPlis replied:
if (!currencyCode.contains(c.toUpperCase())) 

обычно если сравнивают строки то делают через игноркейс а не приводят к регистру

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/config/CurrencyApiConfig.java:
```diff
@@ -1,31 +1,36 @@
-@Component
+@Configuration
+@ConfigurationProperties(prefix = "currency.api")
 public class CurrencyApiConfig {
```
попробуй использовать record из 17 джавы возможно сработает, сам не пробывал, расскажешь мне как попробуешь

AndreyPlis replied:
проперти и конфигурацию разносят по отдельности в разные классы

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/constant/ApiMessage.java:
```diff
@@ -1,11 +1,11 @@
     private String message;
 
-    APIMessageEnum(String message) {
+    ApiMessage(String message) {
```
Не понятно имя энама  что за апи мессадж хочется конкретики, типа код ответа от апи или еще что то.

AndreyPlis replied:
Выражай свои намерения в коде максимально понятно и явно, так как будто тот кто читает впервые это видит и сразу поймет

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/dto/ConvertParametersDto.java:
```diff
@@ -0,0 +1,41 @@
+package ru.svistunovaleksei.tg.currencyconverter.currencyapi.dto;
+
+public class ConvertParametersDto {
```
име класса не понятно. это дто а дто не должно иметь глагола в имени. Только существительное и должно отображать сущность

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/dto/FromToCurrency.java:
```diff
@@ -1,4 +1,4 @@
+package ru.svistunovaleksei.tg.currencyconverter.currencyapi.dto;
 
 import java.util.Map;
 
```
пробуй рекорды почитай про них и попробую юзать

RUMirNul replied:
Попробовал, но как-то с lombok лучше.
1 Геттеры и сеттеры привычней
2 Мне не понравилась запись с параметрами jsonproperty будет как-то так public record ToCurrencyConvertDto(@JsonProperty("currency_name") String currencyName, @JsonProperty("rate") String rate, @JsonProperty("rate_for_amount") String rateForAmount)
3 По мне лучше практика с lombok тк я думаю мало где в продакшене 16+ джава

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/service/AllCurrencyService.java:
```diff
@@ -1,44 +1,34 @@
-        return allCurrenciesNames;
-    }
+    public AllCurrencyDto getAllCurrency() {
+        AllCurrencyDto allCurrencyDto = WebClient.builder()
```
вебклиент не нужно создавать каждый раз.
WebClient.builder()
                .baseUrl(currencyApiConfig.getAllPath())
                .build() явно можно сдлеать раз

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/service/AllCurrencyService.java:
```diff
@@ -1,44 +1,34 @@
+                .get()
+                .retrieve()
+                .bodyToMono(AllCurrencyDto.class)
+                .onErrorReturn(new AllCurrencyDto())
```
лучше явно обработать ошибку и кинуть эксепшен с понятным сообщением и да также обработай если запрос отвалится по таймауту

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/service/AllCurrencyService.java:
```diff
@@ -1,44 +1,34 @@
-            this.allCurrenciesNames.put("RUB", "Российский рубль");
-        }
-        this.status = allCurrency.getStatus();
+        if (!allCurrencyDto.getStatus().equalsIgnoreCase(ApiMessage.SUCCESS.getMessage())) return new AllCurrencyDto();
```
выглядит так что ошибка проглочена

RUMirNul replied:
Это было исправлено в последнем пуше, когда делал обработку ошибок и бросался ServiceUnavailableException(); 
Но обработка ошибок будет переделана

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/currencyapi/service/FromToCurrencyService.java:
```diff
@@ -1,40 +1,38 @@
+        urlParams.put("from", parameters.getFrom());
+        urlParams.put("to", parameters.getTo());
+
+        return WebClient.builder()
```
аналогичные пробелемы клиента как и выше

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/tgbot/handler/MessageHandler.java:
```diff
@@ -168,4 +149,52 @@ private String convertInputData(String input) {
+    private String generateConvertCurrenciesMessage(FromToCurrency fromToCurrency, ConvertParametersDto parameters) {
+        Map<String, ToCurrencyConvert> rates = fromToCurrency.getRates();
+
+        if (!(rates == null || rates.isEmpty())) {
```
коллекия не должна быть нулл

RUMirNul replied:
Проверку убрал, они не нужна была тк мапа не может быть null тк есть проверка на это в FromToCurrencyService

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/tgbot/handler/MessageHandler.java:
```diff
@@ -168,4 +149,52 @@ private String convertInputData(String input) {
+    private String generateAllCurrenciesMessage(AllCurrencyDto allCurrencyDto) {
+        Map<String, String> allCurrencyMap = allCurrencyDto.getCurrencies();
+
+        if (!(allCurrencyMap == null || allCurrencyMap.isEmpty())) {
```
колекция не должна быть нулл

RUMirNul replied:
Проверку убрал, они не нужна была тк мапа не может быть null тк есть проверка на это в AllCurrencyService

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/tgbot/handler/MessageHandler.java:
```diff
@@ -168,4 +149,52 @@ private String convertInputData(String input) {
+        if (!(allCurrencyMap == null || allCurrencyMap.isEmpty())) {
+
+            StringBuilder sb = new StringBuilder();
+            sb.append("*Доступные валюты*\n\n");
```
имеет смысл обядинить в одну строку юзая фичич 17 джавы с текстом

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/tgbot/handler/MessageHandler.java:
```diff
@@ -168,4 +149,52 @@ private String convertInputData(String input) {
+            sb.append("*Доступные валюты*\n\n");
+            sb.append("*Код валюты (Полное назание валюты)*\n");
+
+            for (String key : allCurrencyMap.keySet()) {
```
хм выглядит так что складыванию в одну строку всех валют можно сделать один раз и отдвать готовую строку чем на каждый запрос делать эту работу

RUMirNul replied:
Лучше через кеш, или создать переменную в классе и при первом запросе её заполнять, или вообще в конструкторе вызывать метод, который будет заполнять переменную в классе?

AndreyPlis replied:
не надо в кострукторах писать логику. делаешь спринг кеш прочитай про него

RUMirNul replied:
done

Commit:
Change java version 18 -> 17
Change Spring version 3 -> 2.7

Commit:
Changing the project structure

Commit:
Added tests

Commit:
Added error handling from api

Commit:
Changed bot messages

RUMirNul committed:
Create README.md

Add README.md

AndreyPlis commented on pom.xml:
```diff
@@ -29,6 +29,17 @@
+            <artifactId>spring-boot-starter-webflux</artifactId>
+        </dependency>
+
+        <dependency>
```
зачем эта зависимость?

RUMirNul replied:
Без неё не работает WebClient

Commit:
Changed structure of application.yml
Changed app configuration
Changed CurrencyController to CurrencyService
Removed unnecessary collection checks in MessageHandler
Add new checks for empty collections in FromToCurrencyService and AllCurrencyService

Commit:
Merge remote-tracking branch 'origin/Refactoring' into Refactoring

AndreyPlis commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,12 @@
+    webhookPath:
+
+  currency-api:
+    allCurrencyPath: https://api.getgeoapi.com/v2/currency/list?api_key=APITOKEN
```
не надо хардкодить запросы в ямле
должно быть так
app:
   getgeoapi-url: https://api.getgeoapi.com

остальное дописываешь в коде. так правильнее делать чем как ты написал. потому что если будет в будущем допустим больше запросов к апи ты начнешь еще больше в ямле хардкодить

RUMirNul replied:
Done

AndreyPlis commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,12 @@
+  telegram-bot:
+    name: ConverterCalculatorCurrencyBot
+    token:
+    webhookPath:
```
в ямле не используется камелкейс  должно webhook-path

RUMirNul replied:
Done

AndreyPlis commented on src/test/java/ru/svistunovaleksei/tg/currencyconverter/handler/MessageHandlerTest.java:
```diff
@@ -0,0 +1,225 @@
+    private MessageHandler messageHandler;
+
+    @Test
+    void handler_startMessage() {
```
не надо префикс handler тесты должны начинаться с слова shouldКакое то действие
пример shouldGetAllCurrencies

RUMirNul replied:
Done

AndreyPlis commented on src/test/java/ru/svistunovaleksei/tg/currencyconverter/handler/MessageHandlerTest.java:
```diff
@@ -0,0 +1,225 @@
+        message.setText(TextConstants.help);
+
+        Chat chat = new Chat();
+        chat.setId(13L);
```
волшебные числа в константы

AndreyPlis replied:
а еще лучше не хардкодить id а генерить рандом чтобы данные между тестами не пересекались

RUMirNul replied:
Done

AndreyPlis commented on src/test/java/ru/svistunovaleksei/tg/currencyconverter/handler/MessageHandlerTest.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    @Test
+    void handler_unknownCommandMessage() {
```
в тестах дубляж кода это плохо. логика дублируется меняются только данные, для это используют параметризованные тесты прочитай

RUMirNul replied:
Done

AndreyPlis commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,12 @@
+app:
+  telegram-bot:
+    name: ConverterCalculatorCurrencyBot
+    token:
```
зачем здесь пустые переменные? их не указывают

RUMirNul replied:
там указываются параметры бота, они секретные, поэтому стёр перед пушем

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/TelegramBot.java:
```diff
@@ -1,4 +1,4 @@
-package ru.svistunovaleksei.tg.currencyconverter.tgbot;
+package ru.svistunovaleksei.tg.currencyconverter;
```
лучше не оставлять на верхнем уровне все должно быть по пакетам за исключание main

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/TelegramBotFacade.java:
```diff
@@ -24,6 +24,6 @@ public BotApiMethod<?> handler(Update update) {
             return messageHandler.handler(message);
         }
-        return new SendMessage(message.getChatId().toString(), BotMessageEnum.UNKNOWN_COMMAND_MESSAGE.getMessage());
+        return new SendMessage(message.getChatId().toString(), BotMessage.UNKNOWN_COMMAND_MESSAGE.getMessage());
```
message.getChatId() возможно нулл смотри на подсказки идеи она выделяет проблемный код

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/config/AppConfig.java:
```diff
@@ -25,8 +25,8 @@ public TelegramBot springWebhookBot(TelegramBotFacade telegramBotFacade, SetWebh
 
```
springWebhookBot не надо добавлять префикс спринг

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/controller/WebhookController.java:
```diff
@@ -7,7 +7,7 @@
-import ru.svistunovaleksei.tg.currencyconverter.tgbot.TelegramBot;
+import ru.svistunovaleksei.tg.currencyconverter.TelegramBot;
 
 @RestController
```
@GetMapping
    public ResponseEntity get(){
        return ResponseEntity.ok().build();
    } точно надо?


@PostMapping("/") почему урл просто слеш?

AndreyPlis replied:
просится что то более понятнее чем слеш

RUMirNul replied:
@GetMapping() это пинг-понг, но думаю лучше убрать тк и так будет понятно, что приложение работает
@PostMapping("/") так было сказано в документации, для работы бота на вебхуке, но без указания так же работает
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/dto/ConvertParametersDto.java:
```diff
@@ -0,0 +1,41 @@
+package ru.svistunovaleksei.tg.currencyconverter.dto;
+
+public class ConvertParametersDto {
```
используй либо loombok не надо руками для дто писать всякие сеттеры и тд

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/handler/MessageHandler.java:
```diff
@@ -0,0 +1,206 @@
+public class MessageHandler {
+
+    private ReplyKeyboardMaker replyKeyboardMaker;
+    private CurrencyService currencyService;
```
final

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/handler/MessageHandler.java:
```diff
@@ -0,0 +1,206 @@
+            return sendMessage;
+        }
+
+        FromToCurrencyDto fromToCurrencyDto = null;
```
можно без нулла

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/handler/MessageHandler.java:
```diff
@@ -0,0 +1,206 @@
+            fromToCurrencyDto = currencyService.calculateRateAmount(parameters);
+            sendMessage.setText(generateConvertCurrenciesMessage(fromToCurrencyDto, parameters));
+
+        } catch (InputAmountException e) {
```
не надо плодить эксепшены тем более логика в кетчах одинаковая это прям избыточно, один эксепшен в него передаешь свое сообщение потом когда перехватывешь в кетче достаешь свое сообщение

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/exceptions/IncorrectFromToCurrencyDtoException.java:
```diff
@@ -0,0 +1,4 @@
+package ru.svistunovaleksei.tg.currencyconverter.exceptions;
+
+public class IncorrectFromToCurrencyDtoException extends Exception{
```
много эксепшеном выглядит так что хватит одного

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/handler/MessageHandler.java:
```diff
@@ -0,0 +1,206 @@
+        AllCurrencyDto allCurrencyDto = null;
+        try {
+            allCurrencyDto = currencyService.getAllCurrenciesNames();
+        } catch (ServiceUnavailableException | IncorrectAllCurrencyDtoException e) {
```
ServiceUnavailableException не надо таксать этот эксепшен обработай в сервисе все ошибки и кидай один эксепшен из currencyService. эксепшены в джаве слишком захламляют. не нужно на них строить бизнес логику и перебарщивать с их количеством

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/handler/MessageHandler.java:
```diff
@@ -0,0 +1,206 @@
+        AllCurrencyDto allCurrencyDto = currencyService.getAllCurrenciesNames();
+        Map<String, String> allCurrencyMap = allCurrencyDto.getCurrencies();
+
+        for (Map.Entry<String, String> entry : allCurrencyMap.entrySet()) {
```
цикл замени на  allCurrencyMap.containsValue()

RUMirNul replied:
Он же позвращает boolean, а мне надо получить ключ по значению, где, например, ключ - usd, значение - Доллар США. API возвращает именно "Доллар США", а пользователю я хочу писать "USD"

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/handler/MessageHandler.java:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    private String convertInputData(String input) {
+        String newString = input.replaceAll(", ", ",");
```
зачем? лучше сделать сплит по запятой и потом трим, иначе в твоей логике баг что если после запятой три пробела?

RUMirNul replied:
Исправил с помощью regexp
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/service/AllCurrencyService.java:
```diff
@@ -0,0 +1,46 @@
+
+    public AllCurrencyDto getAllCurrency() throws ServiceUnavailableException, IncorrectAllCurrencyDtoException {
+        AllCurrencyDto allCurrencyDto = WebClient.builder()
+                .baseUrl(currencyApiConfig.getAllCurrencyPath())
```
вотздесь подставляется хост плюс вторая часть урла и да параметры вставляются отдельно

AndreyPlis replied:
параметры урла

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/service/AllCurrencyService.java:
```diff
@@ -0,0 +1,46 @@
+                .get()
+                .retrieve()
+                .bodyToMono(AllCurrencyDto.class)
+                .onErrorReturn(new AllCurrencyDto())
```
нужно обработать ошибку залогировать и кинуть эксепшен свой

AndreyPlis replied:
не надо слать пустые дто когда ошибка, у тебя случилась ошибка а не пришли пустые данные

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/service/AllCurrencyService.java:
```diff
@@ -0,0 +1,46 @@
+                .onErrorReturn(new AllCurrencyDto())
+                .block();
+
+        if (allCurrencyDto == null || allCurrencyDto.getStatus() == null || !allCurrencyDto.getStatus().equalsIgnoreCase(ApiMessage.SUCCESS.getMessage())) {
```
allCurrencyDto == null || allCurrencyDto.getStatus() == null вот это должно уйти

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/service/AllCurrencyService.java:
```diff
@@ -0,0 +1,46 @@
+                .block();
+
+        if (allCurrencyDto == null || allCurrencyDto.getStatus() == null || !allCurrencyDto.getStatus().equalsIgnoreCase(ApiMessage.SUCCESS.getMessage())) {
+            throw new ServiceUnavailableException();
```
сделай один норм эксшепшен

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/service/AllCurrencyService.java:
```diff
@@ -0,0 +1,46 @@
+        }
+    }
+
+    private boolean isValidAllCurrencyDto(AllCurrencyDto allCurrencyDto) {
```
валидация уйдет если обработаешь ошибку, у тебя либо правильно спрасилось вебклиенте либо ошибка веб клиент должен сразу норм возращать результат либо ошибку

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/service/CurrencyService.java:
```diff
@@ -0,0 +1,68 @@
+            if (fromToCurrencyDto.getStatus().equalsIgnoreCase(ApiMessage.SUCCESS.getMessage())) {
+                return fromToCurrencyDto;
+            } else {
+                throw new ServiceUnavailableException();
```
это не надо кидать этот эксепшен имеет другое значение, кидай свой эксепшен

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/service/CurrencyService.java:
```diff
@@ -0,0 +1,68 @@
+            }
+
+        } else {
+            throw new IllegalArgumentException();
```
имеет смысл логировать детально что не так

RUMirNul replied:
Надо будет подключить логгер?

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/service/FromToCurrencyService.java:
```diff
@@ -0,0 +1,55 @@
+        this.currencyApiConfig = currencyApiConfig;
+    }
+
+    public FromToCurrencyDto calculateRateAmount(ConvertParametersDto parameters) throws ServiceUnavailableException, IncorrectFromToCurrencyDtoException {
```
аналогичные проблемы с вебклиентом что и выше сервисом

RUMirNul replied:
Done

AndreyPlis commented on src/main/java/ru/svistunovaleksei/tg/currencyconverter/service/FromToCurrencyService.java:
```diff
@@ -0,0 +1,55 @@
+        urlParams.put("from", parameters.getFrom());
+        urlParams.put("to", parameters.getTo());
+
+        FromToCurrencyDto fromToCurrencyDto = WebClient.builder()
```
вебклиент нужно создавать сразу в конструкторе не надо на кажлдый запрос его создавать также в другом сервисе поправь

RUMirNul replied:
Done

Commit:
Added caching

Commit:
Renamed class
Renamed method name
Edit import for renamed class
Fixed NPE

Commit:
Edit WebhookController

Commit:
Made regexp pattern compile once

Commit:
Add lombok
Redesigned DTO with lombok

Commit:
Deleted redundant exceptions
API requests moved to a general service
Redesigned work with WebClient
Changed application.yml and CurrencyApiConfig.java

Commit:
Fixed an issue when extra space caused an error
Fixed an issue where a comma in the amount of currency being converted returned an incorrect conversion value

Commit:
Fixed warnings

Commit:
Replaced getters and setters on lombok annotations

Commit:
Changed equals on equalsIgnoreCase in MessageHandler.java
Fixed text formatting if there was no text in the message
Renamed tests
Many tests are combined into one parameterized

Commit:
Replaced the standard cache with Caffeine
